plugins/lsp/lspsaga.nix

```diff
@@ -7,220 +7,651 @@
-      warn_sign = notNull cfg.signs.warning;
-      hint_sign = notNull cfg.signs.hint;
-      infor_sign = notNull cfg.signs.info;
-
-      # TODO Fix this!
-      # error_header = notNull cfg.headers.error;
-      # warn_header = notNull cfg.headers.warning;
-      # hint_header = notNull cfg.headers.hint;
-      # infor_header = notNull cfg.headers.info;
-
-      max_diag_msg_width = notNull cfg.maxDialogWidth;
-
-      code_action_icon = notNull cfg.icons.codeAction;
-      finder_definition_icon = notNull cfg.icons.findDefinition;
-      finder_reference_icon = notNull cfg.icons.findReference;
-      definition_preview_icon = notNull cfg.icons.definitionPreview;
-
-      max_finder_preview_lines = notNull cfg.maxFinderPreviewLines;
-
-      rename_prompt_prefix = notNull cfg.renamePromptPrefix;
-
-      border_style = let
-        borderStyle =
-          if cfg.borderStyle == "thin"
-          then 1
-          else if cfg.borderStyle == "rounded"
-          then 2
-          else if cfg.borderStyle == "thick"
-          then 3
-          else null;
-      in
-        borderStyle;
-
-      finder_action_keys = let
-        keys = {
-          open = notNull cfg.keys.finderAction.open;
-          vsplit = notNull cfg.keys.finderAction.vsplit;
-          split = notNull cfg.keys.finderAction.split;
-          quit = notNull cfg.keys.finderAction.quit;
-          scroll_down = notNull cfg.keys.finderAction.scrollDown;
-          scroll_up = notNull cfg.keys.finderAction.scrollUp;
+          virtualText = helpers.defaultNullOpts.mkBool true "Show virtual text at the end of line.";
+
+          priority = helpers.defaultNullOpts.mkInt 100 "Sign priority.";
         };
-      in
-        notEmpty keys;
 
-      code_action_keys = let
-        keys = {
-          quit = notNull cfg.keys.codeAction.quit;
-          exec = notNull cfg.keys.codeAction.exec;
+        beacon = {
+          enable = helpers.defaultNullOpts.mkBool true ''
+            In Lspsaga, some commands jump around in buffer(s).
+            With beacon enabled, it will show a beacon to tell you where the cursor is.
+          '';
+
+          frequency = helpers.defaultNullOpts.mkInt 7 "Frequency.";
         };
-      in
-        notEmpty keys;
-    };
-  in
-    mkIf cfg.enable {
-      extraPlugins = [cfg.package];
+      };
+  };
+
+  config = mkIf cfg.enable {
+    extraPlugins =
+      [cfg.package]
+      ++ (
+        optional
+        (cfg.ui.devicon == null || cfg.ui.devicon)
+        pkgs.vimPlugins.nvim-web-devicons
+      );
 
-      extraConfigLua = ''
-        local saga = require 'lspsaga'
+    warnings =
+      mkIf (
+        # https://nvimdev.github.io/lspsaga/implement/#default-options
+        (isBool cfg.implement.enable && cfg.implement.enable)
+        && (isBool cfg.symbolInWinbar.enable && !cfg.symbolInWinbar.enable)
+      )
+      ["You have enabled the `implement` module but it requires `symbolInWinbar` to be enabled."];
```

Comment:
This is also required for `outline`
https://nvimdev.github.io/lspsaga/outline/

Reply:
Nice catch !

Reply:
Well, as the `outline` module has no `enable` option, I don't think we need an additional warning here.